src/debugging/DockerDebugConfigurationProvider.ts

```diff
@@ -0,0 +1,107 @@
+
+    public provideDebugConfigurations(folder: WorkspaceFolder | undefined, token?: CancellationToken): ProviderResult<DebugConfiguration[]> {
+        return undefined;
+    }
+
+    public resolveDebugConfiguration(folder: WorkspaceFolder | undefined, debugConfiguration: DockerDebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration | undefined> {
+        const debugPlatform = getPlatform(debugConfiguration);
+        return callWithTelemetryAndErrorHandling(
+            `docker-launch/${debugPlatform || 'unknown'}`,
+            async (actionContext: IActionContext) => await this.resolveDebugConfigurationInternal(
+                {
+                    folder: folder || await quickPickWorkspaceFolder('To debug with Docker you must first open a folder or workspace in VS Code.'),
+                    platform: debugPlatform,
+                    actionContext: actionContext,
+                    cancellationToken: token,
+                },
+                debugConfiguration));
+    }
+
+    private async resolveDebugConfigurationInternal(context: DockerDebugContext, originalConfiguration: DockerDebugConfiguration): Promise<DockerDebugConfiguration | undefined> {
+        context.actionContext.telemetry.properties.platform = context.platform;
+
+        const helper = this.getHelper(context.platform);
+        const resolvedConfiguration = await helper.resolveDebugConfiguration(context, originalConfiguration);
+
+        if (resolvedConfiguration) {
+            await this.validateResolvedConfiguration(resolvedConfiguration);
+            await this.registerRemoveContainerAfterDebugging(resolvedConfiguration);
+        }
+
+        // TODO: addDockerSettingsToEnv?
+        return resolvedConfiguration;
+    }
+
+    private async validateResolvedConfiguration(resolvedConfiguration: ResolvedDebugConfiguration): Promise<void> {
+        if (!resolvedConfiguration.type) {
+            throw new Error('No debug type was resolved.');
+        } else if (!resolvedConfiguration.request) {
+            throw new Error('No debug request was resolved.');
+        }
+    }
+
+    private async registerRemoveContainerAfterDebugging(resolvedConfiguration: ResolvedDebugConfiguration): Promise<void> {
+        if (resolvedConfiguration.dockerOptions !== undefined
+            && (resolvedConfiguration.dockerOptions.removeContainerAfterDebug === undefined || resolvedConfiguration.dockerOptions.removeContainerAfterDebug)
+            && resolvedConfiguration.dockerOptions.containerNameToKill !== undefined) {
+            const runningContainerName =
+                (await this.dockerClient.listContainers({ format: '{{.Names}}' }))
+                    .split('\n')
+                    .find(name => name.toLowerCase() === resolvedConfiguration.dockerOptions.containerNameToKill.toLowerCase());
+
+            if (runningContainerName) {
+                await this.dockerClient.removeContainer(runningContainerName, { force: true });

```

Comment:
>removeContainer [](start = 40, length = 15)

It's non-obvious to me why we do the removeContainer call and then register a callback to do the removeContainer call.  Perhaps a comment?

Reply:
Agreed a comment would be good here. The reason is that, in the normal flow, we remove the container after debugging. But, if for some reason that doesn't happen, we can try removing in advance. One such example of the latter that we ran into is if you kill the experimental instance while it is debugging an app, it won't shut down the container, so it is left there.